server/src/main/java/org/elasticsearch/cluster/AckedClusterStateUpdateTask.java

```diff
@@ -64,8 +65,10 @@ public void onAllNodesAcked(@Nullable Exception e) {
         listener.onResponse(newResponse(e == null));
     }
 
-    protected abstract Response newResponse(boolean acknowledged);
-
+    @SuppressWarnings("unchecked")
+    protected Response newResponse(boolean acknowledged) {
+        return (Response) AcknowledgedResponse.of(acknowledged);
```

Comment:
I don't think this cast should be necessary. If we are saying all response classes for AckedClusterStateUpdateTask should be an AcknowledgedResponse, then why is it a generic? Should the generic be extending AcknowledgedResponse?

Reply:
I tried this but annoyingly `org.elasticsearch.xpack.ilm.action.TransportRemoveIndexLifecyclePolicyAction` uses a `Response` class that doesn't extend `AcknowledgedResponse` which prevents us from doing this.
At least I couldn't find a quick+dry way around the problem for now other than what I had before with subclassing `AckedClusterStateUpdateTask` but even that was a lot of extra code, the current version seems simplest to me because changing the response in the ILM action isn't so trivial BwC wise?

Reply:
Is that ILM class actually utilizing the "acked" part of AckedClusterStateUpdateTask then? It seems odd based on naming alone that there is nothing asked about the response?

Reply:
Actually it isn't using any of the acked mechanics so I turned it into a normal update in https://github.com/elastic/elasticsearch/pull/63646/commits/926442ce81b97820b57fbe5712e645a3ab64275e and then constrained the type of the response some to actually by an acknowledged response.

Still need the cast though unfortunately, we can only get around that by sub-classing for plain `AcknowledgedResponse` I think (this is because the listener might have a more constrained type than `AcknowledgedResponse` so without the cast the generic type constraints aren't met).

Reply:
I only see two class that concretely implement AckedClusterStateUpdateTask. Yet they seem to override everything. What benefit does the generic type give us? It seems the abstract class could concretely use AcknowledgedResponse, and those two subclasses can override and refine the response type of newResponse.

Reply:
Fair point, the generic is pointless. I removed it in https://github.com/elastic/elasticsearch/pull/63646/commits/8ca8862564b95e2facc354970913bec2e4e75027 . Still requires one annoying case in the constructor, but much nicer IMO.

We do have a few more than 2 overrides to the response type though in the anonymous implementations. Some of those look completely redundant (e.g. `org.elasticsearch.xpack.core.ilm.action.DeleteLifecycleAction.Response` though and could be removed in a follow-up).

Let me know what you think :)